main.go

```diff
@@ -0,0 +1,186 @@
+		fmt.Println("no port specified in the yaml config file, defaulting to 8000")
+		Config.Port = "8000"
+	}
+
+	if len(Config.Repos) == 0 {
+		log.Fatal(errors.New("no git repos were configured in the yaml config file, at least one repo must be listed for the service to query"))
+	}
+
+	http.HandleFunc("/search", func(w http.ResponseWriter, r *http.Request) {
+		var err error
+
+		if r.Method != "GET" {
+			errorResp(w, fmt.Sprintf("incorect request method %s only the GET method is allowed", r.Method))
+			return
+		}
+
+		req := &SearchRequest{}
+		req.SearchTerm = r.URL.Query().Get("q")
+		if len(req.SearchTerm) == 0 {
+			errorResp(w, "no search term was found, query string must have a 'q' parameter which must be at least 1 character long")
+			return
+		}
+		req.User = r.URL.Query().Get("user")
+
+		searchResp, err := search(req)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("search query could not be compleated, %s", err))
+			return
+		}
+
+		respJSON, err := json.Marshal(searchResp)
+		if err != nil {
+			errorResp(w, fmt.Sprintf("could not marshal query response %s", err))
+			return
+		}
+
+		w.Header().Add("Content-Type", "application/json")
+		w.Write(respJSON)
+	})
+
+	fmt.Println("starting service on port :" + Config.Port)
+	log.Fatal(http.ListenAndServe(":"+Config.Port, nil))
+}
+
+type SearchRequest struct {
+	SearchTerm string `json:"SearchTerm"`
+	User       string `json:"User"`
+}
+
+type Result struct {
+	FileURL string `json:"fileURL"`
+	Repo    string `json:"repo"`
+}
+
+type SearchResponse struct {
+	Results []Result
+}
+
+func search(req *SearchRequest) (*SearchResponse, error) {
+
+	url := "https://api.github.com/search/code?q=" + req.SearchTerm + "+"
+
+	// prevent lots of heap reallocations
+	repos := make([]string, 0, len(Config.Repos))
+	for _, repo := range Config.Repos {
+		// if a user was specified filter only by that specific user
+		if repo[:len(req.User)] == req.User && repo[len(req.User)] == '/' {
+			repos = append(repos, "repo:"+repo)
+		}
+	}
+	if len(repos) == 0 {
+		// we check for 0 repo case when we load the config file so this is a filtring error
+		return nil, fmt.Errorf("no repositories were found belonging to the user %s", req.User)
+	}
+	url += strings.Join(repos, "+")
+	if len(url) > 256 {
+		// this is a restriction of the github api
+		return nil, fmt.Errorf("query must be 256 characters or less, calculated query was %s", url)
+	}
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	// these structs are used to mimic the nesting of the response data
+	type Repo struct {
+		FullName string `json:"full_name"`
+	}
+	type Item struct {
+		Url        string `json:"url"`
+		Repository Repo   `json:"repository"`
+	}
+	type Error struct {
+		Msg string `json:"message"`
+	}
+	respStruct := &struct {
+		Errors []Error `json:"errors"`
+		Items  []Item  `json:"items"`
+	}{}
```

Comment:
the struct names should be private

Reply:
https://mholt.github.io/json-to-go/

Reply:
Answer: Placing these inside the function scope prevents them from muddying the package API/ adding unnecessary structs to the package. These structs only really serve a purpose inside this single function and so don't really need to exists outside of it. Having this as 3 different struct's is probably unnecessary though. I've gone ahead and compressed these structs into a single type and have made that type private rather than public.